src/Functions/widthBucket.cpp

```diff
@@ -0,0 +1,246 @@
+    getValue(const ColumnConst * col_const, const typename ColumnVector<TDataType>::Container * col_vec, const size_t index)
+    {
+        if (nullptr != col_const)
+        {
+            return col_const->getValue<TDataType>();
+        }
+        return col_vec->data()[index];
+    }
+
+    template <typename TResultType, typename TCountType>
+    TResultType calculate(const Float64 operand, const Float64 low, const Float64 high, const TCountType count) const
+    {
+        if (count == 0)
+        {
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Last argument (count) for function {} cannot be 0.", getName());
+        }
+
+        if (operand < low || low >= high)
+        {
+            return 0;
+        }
+        else if (operand >= high)
+        {
+            return count + 1;
+        }
+
+        return static_cast<TResultType>(count * ((operand - low) / (high - low)) + 1);
+    }
+
+    template <is_any_of<UInt8, UInt16, UInt32, UInt64> TCountType>
+    ColumnPtr executeForResultType(const ColumnsWithTypeAndName & arguments, size_t input_rows_count) const
+    {
+        using ResultType = typename NumberTraits::Construct<false, false, NumberTraits::nextSize(sizeof(TCountType))>::Type;
+        auto common_type = std::make_shared<DataTypeNumber<Float64>>();
+
+        std::vector<ColumnPtr> casted_columns;
+        casted_columns.reserve(3);
+        for (const auto argument_index : collections::range(0, 3))
+        {
+            casted_columns.push_back(castColumn(arguments[argument_index], common_type));
+        }
+
+        const auto * operands_vec = getDataIfNotNull(checkAndGetColumn<ColumnVector<Float64>>(casted_columns[0].get()));
+        const auto * lows_vec = getDataIfNotNull(checkAndGetColumn<ColumnVector<Float64>>(casted_columns[1].get()));
+        const auto * highs_vec = getDataIfNotNull(checkAndGetColumn<ColumnVector<Float64>>(casted_columns[2].get()));
+        const auto * counts_vec = getDataIfNotNull(checkAndGetColumn<ColumnVector<TCountType>>(arguments[3].column.get()));
+
+        const auto * operands_col_const = checkAndGetColumnConst<ColumnVector<Float64>>(casted_columns[0].get());
+        const auto * lows_col_const = checkAndGetColumnConst<ColumnVector<Float64>>(casted_columns[1].get());
+        const auto * highs_col_const = checkAndGetColumnConst<ColumnVector<Float64>>(casted_columns[2].get());
+        const auto * counts_col_const = checkAndGetColumnConst<ColumnVector<TCountType>>(arguments[3].column.get());
+
+        assert((nullptr != operands_col_const) ^ (nullptr != operands_vec && operands_vec->size() == input_rows_count));
+        assert((nullptr != lows_col_const) ^ (nullptr != lows_vec && lows_vec->size() == input_rows_count));
+        assert((nullptr != highs_col_const) ^ (nullptr != highs_vec && highs_vec->size() == input_rows_count));
+        assert((nullptr != counts_col_const) ^ (nullptr != counts_vec && counts_vec->size() == input_rows_count));
```

Comment:
I think for this case, it is a bit better to throw an exception (if we cannot cast an argument to `ColumnConst` or `ColumnVector`). I think it is better than to crash in case of some external bug.